File: backend/app/api/v1/endpoints/copilot.py
from typing import List
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from app.api.deps import get_current_user, get_db
from app.models.user import User
from app.schemas.copilot import (
    ChatMessageRequest,
    CopilotMessageSchema,
    GeneratedReportSchema,
    RecommendationsResponse,
    ReportRequest,
)
from app.services.ai.conversation_manager import ConversationManager
from app.services.ai.recommendation_engine import RecommendationEngine
from app.services.ai.report_generator import ReportGenerator

router = APIRouter()

@router.get("/{investigation_id}/history", response_model=List[CopilotMessageSchema])
def get_history(
    investigation_id: UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    manager = ConversationManager(db)
    messages = manager.get_history(investigation_id)
    return messages

@router.post("/{investigation_id}/chat", response_model=CopilotMessageSchema)
async def chat(
    investigation_id: UUID,
    request: ChatMessageRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    try:
        manager = ConversationManager(db)
        ai_message = await manager.chat(investigation_id, request.content)
        return ai_message
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.get("/{investigation_id}/recommendations", response_model=RecommendationsResponse)
async def get_recommendations(
    investigation_id: UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    try:
        engine = RecommendationEngine(db)
        recs = await engine.get_recommendations(investigation_id)
        return RecommendationsResponse(recommendations=recs)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))

@router.post("/{investigation_id}/report", response_model=GeneratedReportSchema)
async def generate_report(
    investigation_id: UUID,
    request: ReportRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    try:
        generator = ReportGenerator(db)
        report = await generator.generate_report(investigation_id, request.report_type, current_user.id)
        return report
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))

File: frontend/src/api/copilot.ts
import { apiClient } from "./client";

export interface CopilotMessage {
  id: string;
  role: "USER" | "ASSISTANT" | "SYSTEM";
  content: string;
  evidence_references: any[];
  created_at: string;
}

export interface GeneratedReport {
  id: string;
  report_type: string;
  content: string;
  generated_at: string;
}

export const getCopilotHistory = async (investigationId: string): Promise<CopilotMessage[]> => {
  const { data } = await apiClient.get<CopilotMessage[]>(`/copilot/${investigationId}/history`);
  return data;
};

export const sendCopilotMessage = async (investigationId: string, content: string): Promise<CopilotMessage> => {
  const { data } = await apiClient.post<CopilotMessage>(`/copilot/${investigationId}/chat`, { content });
  return data;
};

export const getRecommendations = async (investigationId: string): Promise<string[]> => {
  const { data } = await apiClient.get<{ recommendations: string[] }>(`/copilot/${investigationId}/recommendations`);
  return data.recommendations;
};

export const generateReport = async (investigationId: string, reportType: string): Promise<GeneratedReport> => {
  const { data } = await apiClient.post<GeneratedReport>(`/copilot/${investigationId}/report`, { report_type: reportType });
  return data;
};
